src/Microsoft.AspNet.OData.Shared/ODataNullValueMessageHandler.cs

```diff
@@ -0,0 +1,118 @@
+                currentIndex--;
+            }
+
+            // Null value properties should be treated in the same way independent of whether the user asked for the
+            // raw value of the property or a specific format, so we skip the $value segment as it can only be
+            // preceded by a property access segment.
+            if (currentIndex >= 0 && segments[currentIndex] is ValueSegment)
+            {
+                currentIndex--;
+            }
+
+            // Protect ourselves against malformed path segments.
+            if (currentIndex < 0)
+            {
+                return null;
+            }
+
+            KeySegment keySegment = segments[currentIndex] as KeySegment;
+            if (keySegment != null)
+            {
+                // Look at the previous segment to decide, but skip any possible sequence of cast segments in 
+                // between.
+                currentIndex--;
+                while (currentIndex >= 0 && segments[currentIndex] is TypeSegment)
+                {
+                    currentIndex--;
+                }
+                if (currentIndex < 0)
+                {
+                    return null;
+                }
+
+                if (segments[currentIndex] is EntitySetSegment)
+                {
+                    // Return 404 if we were trying to retrieve a specific entity from an entity set.
+                    return HttpStatusCode.NotFound;
+                }
+
+                if (segments[currentIndex] is NavigationPropertySegment)
+                {
+                    // Return 204 if we were trying to retrieve a related entity via a navigation property.
+                    return HttpStatusCode.NoContent;
+                }
+
+                return null;
+            }
+
+            PropertySegment propertySegment = segments[currentIndex] as PropertySegment;
+            if (propertySegment != null)
+            {
+                // Return 204 only if the property is single valued (not a collection of values).
+                return GetChangedStatusCodeForProperty(propertySegment);
+            }
+
+            NavigationPropertySegment navigationSegment = segments[currentIndex] as NavigationPropertySegment;
+            if (navigationSegment != null)
+            {
+                // Return 204 only if the navigation property is a single related entity and not a collection
+                // of entities.
+                return GetChangedStatusCodeForNavigationProperty(navigationSegment);
+            }
+
+            SingletonSegment singletonSegment = segments[currentIndex] as SingletonSegment;
+            if (singletonSegment != null)
+            {
+                // Return 404 for a singleton with a null value.
+                return HttpStatusCode.NotFound;
+            }
+
+            return null;
+        }
+
+        private static HttpStatusCode? GetChangedStatusCodeForNavigationProperty(NavigationPropertySegment navigation)
```

Comment:
Just name the function as "GetChangedStatusCode"

Same for Line 112?

Reply:
This is existing code refactored into Shared so it's look new but it's now. That does not answer the question but it is a reason to leave it as-is.